File: src/engines/__tests__/subtraction.test.ts
import { describe, it, expect } from 'vitest';
import { generateSubtractionSteps } from '../subtraction';

describe('generateSubtractionSteps', () => {
  it('handles simple subtraction with no borrowing', () => {
    const steps = generateSubtractionSteps(89, 23);
    const answerSteps = steps.filter(s => s.type === 'answer_digit');
    const borrowSteps = steps.filter(s => s.type === 'borrow');

    expect(borrowSteps).toHaveLength(0);
    expect(answerSteps).toHaveLength(2);
    // Right-to-left: ones (9-3=6), tens (8-2=6)
    expect(answerSteps[0].correctValue).toBe(6);
    expect(answerSteps[1].correctValue).toBe(6);
  });

  it('handles subtraction with borrowing', () => {
    const steps = generateSubtractionSteps(52, 38);
    // 52 - 38 = 14
    // ones: 2 < 8, need to borrow -> 12 - 8 = 4
    // tens: 4 - 3 = 1
    const answerSteps = steps.filter(s => s.type === 'answer_digit');
    const borrowSteps = steps.filter(s => s.type === 'borrow');

    expect(borrowSteps).toHaveLength(1);
    expect(borrowSteps[0].correctValue).toBe(12); // shows the regrouped value
    expect(answerSteps[0].correctValue).toBe(4);   // ones
    expect(answerSteps[1].correctValue).toBe(1);   // tens
  });

  it('handles 783 - 458 = 325', () => {
    const steps = generateSubtractionSteps(783, 458);
    const answerSteps = steps.filter(s => s.type === 'answer_digit');
    const borrowSteps = steps.filter(s => s.type === 'borrow');

    // ones: 3 < 8, borrow -> 13 - 8 = 5
    // tens: 7 (now 7) - 5 = 2... wait, 8-1=7, 7 < 5 is false, so 7-5=2
    // hundreds: 7 - 4 = 3
    expect(borrowSteps).toHaveLength(1);
    expect(answerSteps[0].correctValue).toBe(5); // ones
    expect(answerSteps[1].correctValue).toBe(2); // tens (7-5=2 after borrow reduced 8 to 7)
    expect(answerSteps[2].correctValue).toBe(3); // hundreds
  });

  it('handles subtraction with zero result', () => {
    const steps = generateSubtractionSteps(100, 100);
    const answerSteps = steps.filter(s => s.type === 'answer_digit');
    expect(answerSteps[0].correctValue).toBe(0);
    expect(answerSteps[1].correctValue).toBe(0);
    expect(answerSteps[2].correctValue).toBe(0);
  });

  it('step order: borrow before answer for each column', () => {
    const steps = generateSubtractionSteps(52, 38);
    // ones column: borrow then answer
    expect(steps[0].type).toBe('borrow');
    expect(steps[1].type).toBe('answer_digit');
    expect(steps[1].correctValue).toBe(4);
    // tens column: just answer
    expect(steps[2].type).toBe('answer_digit');
    expect(steps[2].correctValue).toBe(1);
  });
});

File: go-engines/subtraction_test.go
package engines

import "testing"

func TestGenerateSubtractionSteps_SimpleNoBorrow(t *testing.T) {
	steps := GenerateSubtractionSteps(89, 23)
	answerSteps := filterStepsByType(steps, AnswerDigit)
	borrowSteps := filterStepsByType(steps, Borrow)

	if len(borrowSteps) != 0 {
		t.Fatalf("expected 0 borrow steps, got %d", len(borrowSteps))
	}
	if len(answerSteps) != 2 {
		t.Fatalf("expected 2 answer steps, got %d", len(answerSteps))
	}
	if answerSteps[0].CorrectValue != 6 {
		t.Errorf("ones: expected 6, got %d", answerSteps[0].CorrectValue)
	}
	if answerSteps[1].CorrectValue != 6 {
		t.Errorf("tens: expected 6, got %d", answerSteps[1].CorrectValue)
	}
}

func TestGenerateSubtractionSteps_WithBorrowing(t *testing.T) {
	steps := GenerateSubtractionSteps(52, 38)
	answerSteps := filterStepsByType(steps, AnswerDigit)
	borrowSteps := filterStepsByType(steps, Borrow)

	if len(borrowSteps) != 1 {
		t.Fatalf("expected 1 borrow step, got %d", len(borrowSteps))
	}
	if borrowSteps[0].CorrectValue != 12 {
		t.Errorf("borrow: expected 12, got %d", borrowSteps[0].CorrectValue)
	}
	if answerSteps[0].CorrectValue != 4 {
		t.Errorf("ones: expected 4, got %d", answerSteps[0].CorrectValue)
	}
	if answerSteps[1].CorrectValue != 1 {
		t.Errorf("tens: expected 1, got %d", answerSteps[1].CorrectValue)
	}
}

func TestGenerateSubtractionSteps_783Minus458(t *testing.T) {
	steps := GenerateSubtractionSteps(783, 458)
	answerSteps := filterStepsByType(steps, AnswerDigit)
	borrowSteps := filterStepsByType(steps, Borrow)

	if len(borrowSteps) != 1 {
		t.Fatalf("expected 1 borrow step, got %d", len(borrowSteps))
	}
	if answerSteps[0].CorrectValue != 5 {
		t.Errorf("ones: expected 5, got %d", answerSteps[0].CorrectValue)
	}
	if answerSteps[1].CorrectValue != 2 {
		t.Errorf("tens: expected 2, got %d", answerSteps[1].CorrectValue)
	}
	if answerSteps[2].CorrectValue != 3 {
		t.Errorf("hundreds: expected 3, got %d", answerSteps[2].CorrectValue)
	}
}

func TestGenerateSubtractionSteps_ZeroResult(t *testing.T) {
	steps := GenerateSubtractionSteps(100, 100)
	answerSteps := filterStepsByType(steps, AnswerDigit)

	if answerSteps[0].CorrectValue != 0 {
		t.Errorf("ones: expected 0, got %d", answerSteps[0].CorrectValue)
	}
	if answerSteps[1].CorrectValue != 0 {
		t.Errorf("tens: expected 0, got %d", answerSteps[1].CorrectValue)
	}
	if answerSteps[2].CorrectValue != 0 {
		t.Errorf("hundreds: expected 0, got %d", answerSteps[2].CorrectValue)
	}
}

func TestGenerateSubtractionSteps_StepOrder(t *testing.T) {
	steps := GenerateSubtractionSteps(52, 38)

	if steps[0].Type != Borrow {
		t.Errorf("step 0: expected borrow, got %s", steps[0].Type)
	}
	if steps[1].Type != AnswerDigit || steps[1].CorrectValue != 4 {
		t.Errorf("step 1: expected answer_digit=4, got %s=%d", steps[1].Type, steps[1].CorrectValue)
	}
	if steps[2].Type != AnswerDigit || steps[2].CorrectValue != 1 {
		t.Errorf("step 2: expected answer_digit=1, got %s=%d", steps[2].Type, steps[2].CorrectValue)
	}
}
